File: suffix_array_and_lcp_array.py
"""
In computer science, a suffix array is a sorted array of all suffixes of a string.
It is a data structure used in, among others, full text indices, data compression algorithms, and the field of bibliometrics.
In computer science, the longest common prefix array (LCP array) is an auxiliary data structure to the suffix array.
It stores the lengths of the longest common prefixes (LCPs) between all pairs of consecutive suffixes in a sorted suffix array.
"""

import sys

s = sys.stdin.readline().strip()
n = len(s)
sa = [i for i in range(n)]
pos = [ord(s[i]) for i in range(n)] + [0]
d = 1
while True:
    sa.sort(key=lambda x: (pos[x], pos[min(x + d, n)]))
    tmp = [0] * n
    for i in range(n - 1):
        x, y = sa[i], sa[i + 1]
        tmp[i + 1] = tmp[i] + ((pos[x], pos[min(x + d, n)]) < (pos[y], pos[min(y + d, n)]))
    
    for i in range(n):
        pos[sa[i]] = tmp[i]
    
    if tmp[-1] == n - 1:
        break
    
    d *= 2

for i in range(n):
    print(sa[i] + 1, end=' ')
print()

lcp = [0] * n
lcp[0] = 'x'
k = 0
for i in range(n):
    if pos[i] == 0:
        k -= (k > 0)
        continue
    
    j = sa[pos[i] - 1]
    while (max(i, j) + k < n) and (s[i + k] == s[j + k]):
        k += 1
    
    lcp[pos[i]] = k
    k -= (k > 0)

print(*lcp)

#--------------------------------------------------------------------------------

import sys

s = sys.stdin.readline().strip()
n = len(s)
sa = [i for i in range(n)]
pos = [ord(s[i]) for i in range(n)] + [0]
d = 1
while d <= n:
    cnt = [0] * max(n, 200)
    cnt[0] = d
    for i in range(d, n):
        cnt[pos[i]] += 1
    
    for i in range(1, len(cnt)):
        cnt[i] += cnt[i - 1]
    
    idx = [0] * n
    for i in range(n - 1, -1, -1):
        cnt[pos[min(i + d, n)]] -= 1
        idx[cnt[pos[min(i + d, n)]]] = i
    
    cnt = [0] * max(n, 200)
    for i in range(n):
        cnt[pos[i]] += 1
    
    for i in range(1, len(cnt)):
        cnt[i] += cnt[i - 1]
    
    for i in range(n - 1, -1, -1):
        cnt[pos[idx[i]]] -= 1
        sa[cnt[pos[idx[i]]]] = idx[i]
    
    tmp = [0] * n
    for i in range(n - 1):
        x, y = sa[i], sa[i + 1]
        tmp[i + 1] = tmp[i] + ((pos[x], pos[min(x + d, n)]) < (pos[y], pos[min(y + d, n)]))
    
    for i in range(n):
        pos[sa[i]] = tmp[i]
    
    if tmp[-1] == n - 1:
        break
    
    d *= 2

for i in range(n):
    print(sa[i] + 1, end=' ')
print()

lcp = [0] * n
lcp[0] = 'x'
k = 0
for i in range(n):
    if pos[i] == 0:
        k -= (k > 0)
        continue
    
    j = sa[pos[i] - 1]
    while (max(i, j) + k < n) and (s[i + k] == s[j + k]):
        k += 1
    
    lcp[pos[i]] = k
    k -= (k > 0)

print(*lcp)

File: suffix_array_and_lcp_array.cpp
// BOJ 9248 Suffix Array

#include <bits/stdc++.h>
#define sz size()
#define bk back()
#define fi first
#define se second

using namespace std;
typedef long long ll;
typedef pair<int, int> pii;

void suffix_array(string &s, vector<int> &sa, vector<int> &pos) {
    int n = s.sz;
    for (int i = 0; i < n; i++) {
        sa[i] = i;
        pos[i] = s[i];
    }

    vector<int> tmp(n);
    int d = 1;

    auto cmp = [&](int i, int j) {
        if (pos[i] != pos[j])
            return pos[i] < pos[j];

        i += d;
        j += d;
        return (i < n && j < n) ? (pos[i] < pos[j]) : (i > j);
    };

    while (1) {
        sort(sa.begin(), sa.end(), cmp);
        fill(tmp.begin(), tmp.end(), 0);

        for (int i = 0; i < n - 1; i++)
            tmp[i + 1] = tmp[i] + cmp(sa[i], sa[i + 1]);

        for (int i = 0; i < n; i++)
            pos[sa[i]] = tmp[i];

        if (tmp[n - 1] == n - 1)
            break;

        d *= 2;
    }
}

void longest_common_prefix(string &s, vector<int> &sa, vector<int> &pos, vector<int> &lcp) {
    int n = s.sz;
    int k = 0;

    for (int i = 0; i < n; i++) {
        if (!pos[i]) {
            k = max(k - 1, 0);
            continue;
        }

        int j = sa[pos[i] - 1];
        while (max(i, j) + k < n && s[i + k] == s[j + k])
            k++;

        lcp[pos[i]] = k;
        k = max(k - 1, 0);
    }
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);

    string s;
    cin >> s;

    int n = s.sz;
    vector<int> sa(n), pos(n), lcp(n);

    suffix_array(s, sa, pos);
    longest_common_prefix(s, sa, pos, lcp);

    for (int i = 0; i < n; i++)
        cout << sa[i] + 1 << ' ';
    cout << '\n';

    cout << "x ";
    for (int i = 1; i < n; i++)
        cout << lcp[i] << ' ';
}

//--------------------------------------------------------------------------------

#include <bits/stdc++.h>
#define sz size()
#define bk back()
#define fi first
#define se second

using namespace std;
typedef long long ll;
typedef pair<int, int> pii;

void suffix_array(string &s, vector<int> &sa, vector<int> &pos) {
    int n = s.sz;
    for (int i = 0; i < n; i++) {
        sa[i] = i;
        pos[i] = s[i];
    }

    int d = 1;
    int k = max(n, 200);

    vector<int> cnt(k), idx(n), tmp(n);

    while (1) {
        fill(cnt.begin(), cnt.end(), 0);
        fill(idx.begin(), idx.end(), 0);

        cnt[0] = d;
        for (int i = d; i < n; i++)
            cnt[pos[i]]++;

        for (int i = 1; i < k; i++)
            cnt[i] += cnt[i - 1];

        for (int i = n - 1; i >= 0; i--) {
            cnt[pos[min(i + d, n)]]--;
            idx[cnt[pos[min(i + d, n)]]] = i;
        }

        fill(cnt.begin(), cnt.end(), 0);

        for (int i = 0; i < n; i++)
            cnt[pos[i]]++;

        for (int i = 1; i < k; i++)
            cnt[i] += cnt[i - 1];

        for (int i = n - 1; i >= 0; i--) {
            cnt[pos[idx[i]]]--;
            sa[cnt[pos[idx[i]]]] = idx[i];
        }

        fill(tmp.begin(), tmp.end(), 0);

        for (int i = 0; i < n - 1; i++) {
            int x = sa[i];
            int y = sa[i + 1];

            if (pos[x] < pos[y] || (pos[x] == pos[y] && pos[min(x + d, n)] < pos[min(y + d, n)]))
                tmp[i + 1] = tmp[i] + 1;
            else
                tmp[i + 1] = tmp[i];
        }

        for (int i = 0; i < n; i++)
            pos[sa[i]] = tmp[i] + 1;

        if (tmp[n - 1] == n - 1)
            break;

        d *= 2;
    }
}

void longest_common_prefix(string &s, vector<int> &sa, vector<int> &pos, vector<int> &lcp) {
    int n = s.sz;
    int j = 0;
    int z = 0;

    for (int i = 0; i < n; i++) {
        if (pos[i] == 1) {
            z = max(z - 1, 0);
            continue;
        }

        j = sa[pos[i] - 2];
        while (max(i, j) + z < n && s[i + z] == s[j + z])
            z++;

        lcp[pos[i] - 1] = z;
        z = max(z - 1, 0);
    }
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);

    string s;
    cin >> s;

    int n = s.sz;
    vector<int> sa(n), pos(n + 1), lcp(n);

    suffix_array(s, sa, pos);
    longest_common_prefix(s, sa, pos, lcp);

    for (int i = 0; i < n; i++)
        cout << sa[i] + 1 << ' ';
    cout << '\n';

    cout << "x ";
    for (int i = 1; i < n; i++)
        cout << lcp[i] << ' ';
}
